PIM candidate BSR/RP support

Jafaral opened the pull request:
This Brings full BSR and RP candidate support to FRR (rfc 5059)

Replaces #9180 , in coordination with @eqvinox

Jafaral committed:
pimd: Candidate-RP support

Signed-off-by: David Lamparter <equinox@opensourcerouting.org>
Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

Jafaral committed:
pimd: candidate BSR/RP show commands, move under ip pim bsr

Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

Jafaral committed:
doc: new PIM candidate BSR/RP commands

Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

Jafaral committed:
tests: add candidate BSR/RP topotest

Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

Jafaral committed:
pimd: Candidate-BSR support

Signed-off-by: David Lamparter <equinox@opensourcerouting.org>
Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

donaldsharp commented on pimd/pim_bsm.c:
```diff
@@ -1452,6 +1493,361 @@ int pim_bsm_process(struct interface *ifp, pim_sgaddr *sg, uint8_t *buf,
+		return;
+	}
+
+	zlog_debug("Candidate-RP (-, %pPA) deregistering self to %pPA",
```
Let's not have a unguarded debug.  Either put this behind a guard or change the level

donaldsharp commented on pimd/pim_nb_config.c:
```diff
@@ -2524,6 +2524,152 @@ static void yang_addrsel(struct cand_addrsel *addrsel,
+
+		scope->cand_bsr_prio = yang_dnode_get_uint8(args->dnode, NULL);
+
+		CPP_NOTICE("FIXME: force prio update");
```
I'm not a fan of leaving a CPP_NOTICE in that does not have a date guard or something that is going to be fixed emminently.  If this is work that needs to be done then let's do the work now or just remove this notice.  I don't want everyone compiling FRR to live with this for the forseeable future

Jafaral replied:
Must have missed this one. Will fix.

donaldsharp commented:
Ran the test and got this new memory leaks:

```
  ## showing active allocations in memory group pimd
pimd:     PIM BSR C-RP                  :      2 *         96
pimd:     PIM BSR range                 :      3 *         80
pimd:     PIM BSR C-RP range item       :      4 *         88
```

My expectation is that we should not be adding new code that leaks

donaldsharp commented on pimd/pim6_cmd.c:
```diff
@@ -1856,20 +1856,21 @@ DEFPY (show_ipv6_pim_secondary,
 	return pim_show_secondary_helper(vrf, vty);
 }
 
-DEFPY (show_ipv6_pim_cand_rp,
```
Where did this command go?  Where is the backwards compatiblity for this?  Someone is using this command we cannot just remove it.

Jafaral replied:
This is a new command. I added new show command beyond what we have in the first couple of commits in a later commit. I didn't go back to make the change in the original commit to avoid some code churn, but I can do that if you think it is necessary.

donaldsharp commented on pimd/pim6_cmd.c:
```diff
@@ -1913,12 +1914,12 @@ DEFPY (show_ipv6_pim_cand_rp,
 
 DEFPY (show_ipv6_pim_bsr_rpdb,
        show_ipv6_pim_bsr_rpdb_cmd,
-       "show ipv6 pim bsr candidate-rps [vrf VRF_NAME] [json$uj]",
```
again this command is just gone.  We need to provide an upgrade path and cannot just remove commands

Jafaral replied:
same comment above.

donaldsharp commented on pimd/pim_cmd.c:
```diff
@@ -2720,22 +2720,100 @@ DEFPY (show_ip_pim_bsrp,
 	return pim_show_group_rp_mappings_info_helper(vrf, vty, !!json);
 }
 
-DEFUN (show_ip_pim_cand_rp,
```
another command just removed

Jafaral replied:
same comment above.

donaldsharp commented on pimd/pim_cmd.c:
```diff
@@ -2780,33 +2858,32 @@ DEFUN (show_ip_pim_cand_rp,
 	return CMD_SUCCESS;
 }
 
-DEFUN (show_ip_pim_bsr_rpdb,
```
another command just removed?

Jafaral replied:
This is the only command that existed before the new BSR as part of this PR. I made  `show_ip_pim_bsr_rpdb` a hidden command, I didn't remove it. Since now we have multiple bits of BSR info, I chose to move all show commands under a common command `show ip pim bsr...`, just for consistency and better user interface.

Jafaral replied:
To clarify further, we now have 
```
show ip pim bsr rp-info
show ip pim bsr candidate-rp
show ip pim bsr groups
show ip pim bsr
```
and more...

Jafaral commented:
> Ran the test and got this new memory leaks:
> 
> ```
>   ## showing active allocations in memory group pimd
> pimd:     PIM BSR C-RP                  :      2 *         96
> pimd:     PIM BSR range                 :      3 *         80
> pimd:     PIM BSR C-RP range item       :      4 *         88
> ```
> 
> My expectation is that we should not be adding new code that leaks

will fix.

nabahr commented on pimd/pim6_cmd.c:
```diff
@@ -1259,6 +1259,154 @@ DEFPY (no_ipv6_pim_ucast_bsm,
+	if (no)
+		nb_cli_enqueue_change(vty, cand_rp_xpath, NB_OP_DESTROY, NULL);
+	else {
+		char xpath2[XPATH_MAXLEN + 24];
```
Same as before, enqueue with relative paths and apply with a base path.

nabahr commented on lib/command.h:
```diff
@@ -399,6 +399,7 @@ struct cmd_node {
 #define IPV6_STR "IPv6 information\n"
 #define IP_ADDR_STR "IPv4 Address\n"
 #define IP6_ADDR_STR "IPv6 Address\n"
+#define BSR_STR "Bootstrap Router configuration\n"
```
Why is this in the lib/command.h and not in pim_cmd.h?
If this is only going to be used by pim it should be defined there.

nabahr commented on pimd/pim_cmd.c:
```diff
@@ -4376,6 +4560,154 @@ DEFPY_ATTR(no_ip_pim_rp_prefix_list,
+	if (no)
+		nb_cli_enqueue_change(vty, cand_bsr_xpath, NB_OP_DESTROY, NULL);
+	else {
+		char xpath2[XPATH_MAXLEN + 16];
```
Enqueue with relative path, apply with base path

nabahr reviewed (commented):
Just a few minor comments

mobash-rasool commented on pimd/pim6_cmd.c:
```diff
@@ -1719,6 +1867,107 @@ DEFPY (show_ipv6_pim_secondary,
+				    event_timer_remain_msec(
+					    scope->cand_rp_adv_timer));
+
+		vty_out(vty, "%s\n",
```
You can use "vty_json" instead of vty_out and json_object_free()

donaldsharp replied:
Why was this resolved?  I started to write a comment about this as well, I see that vty_json is not being used in several spots

mobash-rasool commented on pimd/pim_bsm.c:
```diff
@@ -212,36 +255,132 @@ static inline void pim_bs_timer_restart(struct bsm_scope *scope, int bs_timeout)
+	event_add_read(router->master, bsm_unicast_sock_read, scope,
+		       scope->unicast_sock, &scope->unicast_read);
+
+	for (i = 0; i < router->packet_process; i++) {
```
Line 273 to Line 320 is same code as in function pim_sock_read, wont it be better if we can make a common function and reuse the code.

Jafaral replied:
done. Added as a separate commit for easy review.

mobash-rasool commented on pimd/pim_msg.h:
```diff
@@ -148,6 +148,7 @@ struct pim_encoded_source_ipv6 {
 typedef struct pim_encoded_ipv4_unicast pim_encoded_unicast;
 typedef struct pim_encoded_group_ipv4   pim_encoded_group;
 typedef struct pim_encoded_source_ipv4  pim_encoded_source;
+#define PIM_MSG_ADDRESS_FAMILY		PIM_MSG_ADDRESS_FAMILY_IPV4
```
We do have this already defined in pim_tlv.c file as shown below:
#if PIM_IPV == 4
#define PIM_MSG_ADDRESS_FAMILY PIM_MSG_ADDRESS_FAMILY_IPV4
#else
#define PIM_MSG_ADDRESS_FAMILY PIM_MSG_ADDRESS_FAMILY_IPV6
#endif

Do you mind deleting those.

Jafaral commented:
Working on addressing comments from @nabahr  and @mobash-rasool

Jafaral committed:
pimd: refactor pim_sock_read to be used with unicast bsr msgs

Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

ton31337 commented on pimd/pim6_cmd.c:
```diff
@@ -1719,6 +1775,107 @@ DEFPY (show_ipv6_pim_secondary,
+			vty_out(vty, "{}\n");
+		else
+			vty_out(vty,
+				"This router is not currently operating as Candidate RP\n");
```
I suggest adding this also to JSON output, e.g.:

```
{"warning": "..."}
```

Jafaral replied:
This isn't a warning really, just informational.

Jafaral replied:
I.e, the correct behavior it return an empty object.

ton31337 replied:
Neither agree, nor disagree, e.g. we have in BGP such patterns:
```c
	if (!peer) {
		if (use_json) {
			json_object *json_no = NULL;
			json_no = json_object_new_object();
			json_object_string_add(json_no, "warning",
					       "No such neighbor in this view/vrf");
			vty_json(vty, json_no);
		} else
			vty_out(vty, "No such neighbor in this view/vrf\n");
		return NULL;
	}
```
or
```c
	if (is_synchronized(rpki_vrf)) {
		if (strmatch(prefixkind, "prefix-count"))
			print_prefix_table(vty, rpki_vrf, json, true);
		else
			print_prefix_table(vty, rpki_vrf, json, false);
	} else {
		if (json) {
			json_object_string_add(json, "error", "No Connection to RPKI cache server.");
			vty_json(vty, json);
		} else
			vty_out(vty, "No connection to RPKI cache server.\n");
		return CMD_WARNING;
	}
```
Anyway, it's not a blocker. Just an improvement.

Jafaral replied:
Both of these are a bit different. The first , if you query a neighbor that doesn't exist, a warning is a good idea about it. The second case is an error, so that does need an error return. So I see why those have them. The case I have is purely informational.

ton31337 commented on pimd/pim6_cmd.c:
```diff
@@ -1719,6 +1775,107 @@ DEFPY (show_ipv6_pim_secondary,
+		return CMD_SUCCESS;
+	}
+
+	vty_out(vty, "Candidate-RP\nAddress:   %pPA\nPriority:  %u\n\n",
```
Comestic change: `"Candidate-RP\nAddress:   ` (double whitespace)...

ton31337 commented on pimd/pim_bsm.c:
```diff
@@ -1452,7 +1599,593 @@ int pim_bsm_process(struct interface *ifp, pim_sgaddr *sg, uint8_t *buf,
+	unsigned int timer = PIM_BS_TIME;
+
+	if (scope->changed_bsm_trigger) {
+		zlog_debug("Sending triggered BSM");
```
Unguarded debug?

ton31337 commented on pimd/pim_bsm.c:
```diff
@@ -1452,7 +1599,593 @@ int pim_bsm_process(struct interface *ifp, pim_sgaddr *sg, uint8_t *buf,
+	scope->current_bsr_prio = scope->cand_bsr_prio;
+	scope->current_bsr = scope->bsr_addrsel.run_addr;
+
+	//pim_bsm_frags_free(scope);
```
Can we drop this?

Jafaral commented:
I addressed all comments on this PR.

mobash-rasool commented on doc/user/pimv6.rst:
```diff
@@ -395,7 +418,24 @@ General multicast routing state
    Display current bsr, its uptime and last received bsm age.
 
-.. clicmd:: show ipv6 pim bsrp-info [vrf NAME] [json]
+.. clicmd:: show ipv6 pim bsr candidate-bsr [vrf NAME] [json]
```
Where is this command defined? I cant find it here.

Jafaral replied:
added, and refactored the show command for v4/v6 while I was at it.

Jafaral committed:
pimd: make sure the bsr message is coming from the neighbor

This change re-adds an additional check bsr rpf that was removed
in 2c6a32f9be3c687fe5c52d055f21912da8364ec6.

Without that check, bsr messages is resent (causing a loop) if we
have more than one pim neighbor on a link.

Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

Jafaral committed:
tests: don't set oif to a list if it is None

Signed-off-by: Jafar Al-Gharaibeh <jafar@atcorp.com>

Jafaral commented:
@greptile review
